extensions/common/istio_dimensions.h

```diff
@@ -0,0 +1,101 @@
+/* Copyright 2019 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <set>
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_replace.h"
+#include "absl/strings/string_view.h"
+
+namespace Wasm {
+namespace Common {
+namespace {
+// This map maintains the name of Istio dimension and their key to
+// IstioDimensions::dimensions map.
+absl::flat_hash_map<const absl::string_view, unsigned int> dimensions_name_map;
+}  // namespace
+
+// A structure that can hold multiple Istio dimensions(metadata variables).
+// This could be use to key caches based on Istio dimensions for various
+// filters.
+// TODO: Add support for evaluating dynamic Istio dimensions.
+struct IstioDimensions {
+ public:
+  // This is the map whose key is the assigned integer value for the dimension
+  // in dimensions_name_map and value is the runtime value of the dimension.
+  absl::flat_hash_map<unsigned int, std::string> dimensions;
+
+  static void InitIstioDimensions(
+      std::vector<const absl::string_view> dimensions_name_set) {
+    // TODO: decide if we should clear dimensions cache too.
+    dimensions_name_map.clear();
+    unsigned int i = 0;
+    for (const auto& elem : dimensions_name_set) {
+      dimensions_name_map[elem] = i++;
+    }
+  }
+
+  // maps property with it's value.
+  void map(const absl::string_view name, std::string value) {
+    dimensions[dimensions_name_map[name]] = value;
```

Comment:
Why not make use of std::tuple? It's going to be much faster than double map lookup.

Reply:
I did double map lookup to reduce memory footprint... am hoping dimensions name map would be really small.. something like 1-25 entries, which will basically contain name of variable as a string, hence making lookup faster.

And IstioDimensions which would actually be a key to the cache just has a numeric name along with string value as compared to a pair of string name and string value making it less in size...

Reply:
I think the real question is whether all the keys are known at compile time or not. If they are, then a struct should be faster (maps probably take more memory unless they have compiler optimizations to specialize them to structs). If we don't know all the keys, then map does make sense.

Reply:
Agreed, one of the cases I had in my mind was supporting dynamic expressions and keying based on that, so this cache can be used to sample say on path or source...

Reply:
Thoughts @mandarjog @douglas-reid @bianpengyuan ?

Reply:
I think we can start with a struct. Even for dynamic expressions, we will use one that evaluates to a string right? I think it would still be cheaper.

Reply:
How about we change the dimensions map to a vector, doesn't really need it to be a map.. we can index into it based on dimensions_name_map value.

Reason, am thinking of this is that say if we share this cache with stats filter, then we would end up declaring 20 variables, and even when mostly empty for logs filter, they would still occupy extra 18 bytes in the key of the logs filter cache. The size of maintaining dimensions_name_map in this case would start paying back after around 25-30 entries..

Reply:
Talked to @mandarjog offline and changes to struct, the way it's in stats plugin